packages/langium/src/lsp/folding-range-provider.ts

```diff
@@ -0,0 +1,120 @@
+                            treeIterator.prune();
+                        }
+                    }
+                } while (!result.done);
+            }
+
+            this.collectCommentFolding(document, root, acceptor);
+        }
+    }
+
+    /**
+     * Template method to determine whether the specified `AstNode` should be handled by the folding range provider.
+     * Returns true by default for all nodes. Returning false only ignores the specified node and not its content.
+     * To ignore the content of a node use `shouldProcessContent`.
+     */
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    protected shouldProcess(node: AstNode): boolean {
+        return true;
+    }
+
+    /**
+     * Template method to determine whether the content/children of the specified `AstNode` should be handled by the folding range provider.
+     * Returns true by default for all nodes. Returning false ignores _all_ content of this node, even transitive ones.
+     * For more precise control over foldings use the `shouldProcess` method.
+     */
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    protected shouldProcessContent(node: AstNode): boolean {
+        return true;
+    }
+
+    protected collectObjectFolding(document: LangiumDocument, node: AstNode, acceptor: FoldingRangeAcceptor): void {
+        const cstNode = node.$cstNode;
+        if (cstNode) {
+            const foldingRange = this.toFoldingRange(document, cstNode);
+            if (foldingRange) {
+                acceptor(foldingRange);
+            }
+        }
+    }
+
+    protected collectCommentFolding(document: LangiumDocument, node: AstNode, acceptor: FoldingRangeAcceptor): void {
+        const cstNode = node.$cstNode;
+        if (cstNode) {
+            for (const node of streamCst(cstNode)) {
+                if (node instanceof LeafCstNodeImpl && this.commentNames.includes(node.tokenType.name)) {
+                    const foldingRange = this.toFoldingRange(document, node, FoldingRangeKind.Comment);
+                    if (foldingRange) {
+                        acceptor(foldingRange);
+                    }
+                }
+            }
+        }
+    }
+
+    protected readonly letterTestRegex = /\w/;
+
+    protected toFoldingRange(document: LangiumDocument, node: CstNode, kind?: string): FoldingRange | undefined {
+        const { start, end } = node.range;
+        const startPosition = document.positionAt(start);
+        let endPosition = document.positionAt(end);
+        // Don't generate foldings for nodes that are less than 3 lines
+        if (endPosition.line - startPosition.line < 2) {
+            return undefined;
+        }
+        // As we don't want to hide the end token like 'if { ... --> } <--',
+        // we simply select the end of the previous line as the end position
+        endPosition = document.positionAt(document.offsetAt({ line: endPosition.line, character: 0 }) - 1);
```

Comment:
This is only useful if the end token is actually a closing brace. For grammar rules, for example, it would be better to hide the whole definition. Can this be solved here generically, or do we need a customization for the grammar language?

Reply:
I would argue that the default behavior should not include the last line of the object in the folding. This is basically how every programming language does it:

![image](https://user-images.githubusercontent.com/4377073/132305010-a0330bcb-1d6a-4f17-852c-7e430a4fea24.png)

But yes, it makes sense for the grammar language. I'll implement a template method for this.